Feature/kak/compare page#253

flibbertigibbet opened the pull request:
## Overview

Adds neighborhood comparison page.


### Demo

![image](https://cloud.githubusercontent.com/assets/960264/25393366/48a7bd6e-29a9-11e7-8839-ff969b59be9f.png)



### Notes

The styling for the 'remove' button at the top of each column appears to be broken; it was in the DOM, but not visible. Added button classes to get it to show up, but the icon is missing and there's no text.

Does not implement the list page selection/navigation, which is #251.

Does not implement the map thumbnail card, which needs #337.

Does not implement the footer cards with the metric descriptions (have no descriptions to display yet).

## Testing Instructions

 * Navigate there directly with one or more neighborhood UUIDs: http://localhost:9301/#/compare///bab4bbb5-d209-49be-a577-0b8ecb00f9bd


Closes #253

KlaasH commented on src/angularjs/src/app/compare/compare.controller.js:
```diff
@@ -0,0 +1,84 @@
+            } else {
+                path += $stateParams.place1 + '/' + $stateParams.place2 + '/';
+            }
+            $location.path(path);
```
I think this would be better as a state transition rather than a `$location` update.  I believe
```
var newParams = _.extend({}, $stateParams);
newParams['place' + (num + 1)] = '';
$state.go('compare', newParams);
```
would do it.

flibbertigibbet replied:
That's so much better.

KlaasH commented on src/angularjs/src/app/compare/compare.controller.js:
```diff
@@ -0,0 +1,84 @@
+            ctl.places[num] = {};
+
+            Neighborhood.query({uuid: uuid}).$promise.then(function(data) {
+                ctl.places[num].neighborhood = new Neighborhood(data);
```
Filling the object in pieces makes it render in chunks.  I think it would be better to assign the whole thing to the controller at once.  (Switching to a shared service that returns the whole `place` would also imply setting it on the controller all at once.)

flibbertigibbet replied:
I agree it looks odd to see the columns load separately, split seconds apart. Then again, it seems to generally be preferred web design to present content sooner, piecemeal, than to leave the user looking at an empty page. @CloudNiner maybe have thoughts on how best to load here?

CloudNiner replied:
Hmm, I don't feel strongly about this one either way. I might lean slightly towards displaying all three at the same time for two reasons:
1. Each individual piece of content is less valuable until all the content is presented
2. Loading each block individually makes the page load as a whole feel a bit janky as content pops in piecemeal

If one of the blocks fails to load, you could still show the remaining blocks, and have some type of error block in place of the ones that failed to load. I think that would look the same in either case.

Not to punt down the chain, but @designmatty may have a preference as this feels more of a UI/UX issue.

designmatty replied:
This is definitely one of those very edge-case situations which doesn't have a right or wrong solution. I think I would prefer to show them once all have loaded, and we could potentially show the user some loading spinner while they are loading.

flibbertigibbet commented:
Okay, moved module under places (URL now like http://localhost:9301/#/places/compare//bab4bbb5-d209-49be-a577-0b8ecb00f9bd//). Delay results display until all loaded. Also added thumbnail map, since that's gotten merged.

KlaasH commented on src/angularjs/src/app/places/compare/compare.controller.js:
```diff
@@ -0,0 +1,94 @@
+
+            // do not display any place until all places have been retrieved
+            $q.all(promises).then(function(results) {
+                ctl.places = results;
```
It was mainly the individual cards loading in chunks that looked bad to me, so loading each place into `ctl.places[offset]` as it arrives would be fine as well, I think.  Not sure which is better, but the difference is minor and this is here, so this is good.

flibbertigibbet replied:
Delaying display in any case would require knowing when all promises have either returned or failed, which as far as I can see would still require a promises-aware refactor, which is why I went with this.
